README.md
# Prometheus exporter for HHVM

Export health and runtime metrics from HHVM admin interface into
[Prometheus](https://prometheus.io) metrics.

## Running

The easiest way to run `hhvm_exporter` is via a virtualenv:

  virtualenv .venv
  .venv/bin/python setup.py develop
  .venv/bin/hhvm_exporter

By default metrics are polled from `http://localhost:9002` and exposed on TCP port `9192`:

  curl localhost:9192/metrics -s | grep -v '^#'
  hhvm_load 0.0
  hhvm_memory_strings_bytes 13641543.0
  hhvm_process_memory_bytes{segment="rss"} 443691008.0
  hhvm_process_memory_bytes{segment="shared"} 101527552.0
  ...
  hhvm_startup 1477007598.0
  hhvm_up 1.0
  hhvm_scrape_duration_seconds_count 14.0
  hhvm_scrape_duration_seconds_sum 0.0016635158681310713


debian/changelog
prometheus-hhvm-exporter (0.4-1) unstable; urgency=medium

  * Export APC metrics

 -- Filippo Giunchedi <filippo@debian.org>  Thu, 19 Oct 2017 10:59:02 +0200

prometheus-hhvm-exporter (0.3-1) unstable; urgency=low

  * Raise default logging level to WARNING

 -- Filippo Giunchedi <filippo@wikimedia.org>  Sat, 20 May 2017 12:41:43 +0200

prometheus-hhvm-exporter (0.2-1) unstable; urgency=low

  * Fix duplicate 'hhvm_tc_used_bytes'

 -- Filippo Giunchedi <filippo@wikimedia.org>  Thu, 15 Dec 2016 01:38:36 +0000

prometheus-hhvm-exporter (0.1-1) unstable; urgency=low

  * Initial packaging

 -- Filippo Giunchedi <filippo@wikimedia.org>  Fri, 18 Nov 2016 00:53:36 +0000


hhvm_exporter_test.py
# Copyright 2016 Filippo Giunchedi
#                Wikimedia Foundation
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


import unittest

from hhvm_exporter import HHVMCollector


class TestHHVMCollector(unittest.TestCase):
    def setUp(self):
        self.c = HHVMCollector('http://localhost:9090')

    def testCollectStatus(self):
        expected_names = ['hhvm_startup', 'hhvm_build_info']
        self.assertMetricEqual(
            [x.name for x in self.c._collect_status({})], expected_names)

    def testCollectMemory(self):
        expected_names = [
            'hhvm_memory_strings_count', 'hhvm_process_memory_bytes',
            'hhvm_memory_strings_bytes', 'hhvm_memory_success',
            'hhvm_tc_used_bytes', 'hhvm_process_memory_size_bytes']
        self.assertMetricEqual(
            [x.name for x in self.c._collect_memory({})], expected_names)

    def testCollectHealth(self):
        expected_names = [
            'hhvm_funcs_total', 'hhvm_hhbc_ro_arena_capacity_bytes',
            'hhvm_load', 'hhvm_queued', 'hhvm_rds_local_bytes',
            'hhvm_rds_persistent_bytes', 'hhvm_rds_used_bytes',
            'hhvm_tc_used_bytes', 'hhvm_units_total']
        self.assertMetricEqual(
            [x.name for x in self.c._collect_health({})], expected_names)

    def assertMetricEqual(self, actual, expected):
        return self.assertListEqual(sorted(actual), sorted(expected))


test/test_collector.py
# Copyright 2016 Filippo Giunchedi
#                Wikimedia Foundation
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


import unittest

from hhvm_exporter.exporter import HHVMCollector


class TestHHVMCollector(unittest.TestCase):
    def setUp(self):
        self.c = HHVMCollector('http://localhost:9090')

    def testCollectStatus(self):
        expected_names = ['hhvm_startup', 'hhvm_build_info']
        self.assertMetricEqual(
            [x.name for x in self.c._collect_status({})], expected_names)

    def testCollectMemory(self):
        expected_names = [
            'hhvm_memory_strings_count', 'hhvm_process_memory_bytes',
            'hhvm_memory_strings_bytes', 'hhvm_memory_success',
            'hhvm_process_memory_size_bytes']
        self.assertMetricEqual(
            [x.name for x in self.c._collect_memory({})], expected_names)

    def testCollectHealth(self):
        expected_names = [
            'hhvm_funcs_total', 'hhvm_hhbc_ro_arena_capacity_bytes',
            'hhvm_load', 'hhvm_queued', 'hhvm_rds_local_bytes',
            'hhvm_rds_persistent_bytes', 'hhvm_rds_used_bytes',
            'hhvm_tc_used_bytes', 'hhvm_units_total']
        self.assertMetricEqual(
            [x.name for x in self.c._collect_health({})], expected_names)

    def testCollectAPC(self):
        expected_names = [
            'hhvm_apc_value_bytes', 'hhvm_apc_key_bytes',
            'hhvm_apc_mmap_bytes', 'hhvm_apc_memory_primed_bytes',
            'hhvm_apc_key_count', 'hhvm_apc_key_primed_count',
            'hhvm_apc_key_primed_memory_count',
            'hhvm_apc_pending_delete_bytes']
        self.assertMetricEqual(
            [x.name for x in self.c._collect_apc('')], expected_names)

    def assertMetricEqual(self, actual, expected):
        return self.assertListEqual(sorted(actual), sorted(expected))


hhvm_exporter/exporter.py
#!/usr/bin/python
# Copyright 2016 Filippo Giunchedi
#                Wikimedia Foundation
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import argparse
import datetime
import logging
import re
import sys
import time

import requests

from prometheus_client import start_http_server, Summary
from prometheus_client.core import (CounterMetricFamily, GaugeMetricFamily,
                                    REGISTRY)

log = logging.getLogger(__name__)


class HHVMCollector(object):
    scrape_duration = Summary(
            'hhvm_scrape_duration_seconds', 'HHVM scrape duration')

    def __init__(self, admin_url):
        self.url = admin_url

    def _fetch_json(self, url):
        response = self._fetch_url(url)
        if response is not None:
            return response.json()
        return response

    def _fetch_text(self, url):
        response = self._fetch_url(url)
        if response is not None:
            return response.text
        return response

    def _fetch_url(self, url):
        try:
            response = requests.get(url, timeout=2)
            response.raise_for_status()
            return response
        except (requests.ConnectionError,
                requests.Timeout,
                requests.HTTPError,
                ValueError):
            return None

    @scrape_duration.time()
    def collect(self):
        health_data = self._fetch_json(self.url + '/check-health')
        for metric in self._collect_health(health_data):
            yield metric

        memory_data = self._fetch_json(self.url + '/memory.json')
        for metric in self._collect_memory(memory_data):
            yield metric

        status_data = self._fetch_json(self.url + '/status.json')
        for metric in self._collect_status(status_data):
            yield metric

        apc_info = self._fetch_text(self.url + '/dump-apc-info')
        for metric in self._collect_apc(apc_info):
            yield metric

        up = GaugeMetricFamily('hhvm_up', 'HHVM admin interface is up')
        if not all([status_data, memory_data, health_data]):
            up.add_metric([], 0)
        else:
            up.add_metric([], 1)

        yield up

    def _collect_apc(self, data):
        if data is None:
            raise StopIteration()

        metrics = {
            'value_size': GaugeMetricFamily('hhvm_apc_value_bytes',
                                            'Memory usage from all live values'),
            'key_size': GaugeMetricFamily('hhvm_apc_key_bytes',
                                          'Memory usage from all keys'),
            'mapped_to_file_data_size': GaugeMetricFamily(
                                        'hhvm_apc_mmap_bytes',
                                        'Bytes stored in the paged out memory file'),
            'in_memory_primed_data_size': GaugeMetricFamily(
                                        'hhvm_apc_memory_primed_bytes',
                                        'Bytes that were primed and went into memory'),

            'entries_count': GaugeMetricFamily('hhvm_apc_key_count',
                                               'Number of keys (entries)'),
            'primed_entries_count': GaugeMetricFamily('hhvm_apc_key_primed_count',
                                                      'Number of primed keys'),
            'in_memory_primed_entries_count': GaugeMetricFamily(
                                              'hhvm_apc_key_primed_memory_count',
                                              'Number of primed keys live (in memory)'),
            'pending_deletes_via_treadmill_size': GaugeMetricFamily(
                                                  'hhvm_apc_pending_delete_bytes',
                                                  'Bytes pending deletion via treadmill'),
        }

        for line in data.splitlines():
            if ':' not in line:
                continue
            key, value = re.split(': *', line, 1)
            metric_name = key.lower().replace(' ', '_')

            metric_family = metrics.get(metric_name)
            if metric_name is None:
                log.warn('Unknown metric %r from line %r', metric_name, line)
                continue

            try:
                value = float(value)
            except ValueError:
                value = float('nan')
            metric_family.add_metric([], value)

        for metric in metrics.values():
            yield metric

    def _collect_health(self, data):
        if data is None:
            raise StopIteration()

        metrics = {
           'load': GaugeMetricFamily(
                   'hhvm_load',
                   'Number of threads actively servicing requests'),
           'queued': GaugeMetricFamily(
                   'hhvm_queued', 'Number of queued jobs'),
           'hhbc-roarena-capac': GaugeMetricFamily(
                   'hhvm_hhbc_ro_arena_capacity_bytes',
                   'Read-only arena capacity, used only in RepoAuth mode'),
           'rds': GaugeMetricFamily(
                   'hhvm_rds_used_bytes', 'RDS total bytes usage'),
           'rds-local': GaugeMetricFamily(
                   'hhvm_rds_local_bytes', 'RDS local region bytes usage'),
           'rds-persistent': GaugeMetricFamily(
                   'hhvm_rds_persistent_bytes',
                   'RDS persistent region bytes usage'),
           'units': GaugeMetricFamily(
                   'hhvm_units_total', 'Number of loaded units'),
           'funcs': GaugeMetricFamily(
                   'hhvm_funcs_total', 'Number of functions created'),
        }

        for metric_name in metrics.keys():
            metrics[metric_name].add_metric(
                    [], data.get(metric_name, float('nan')))

        metrics['tc'] = GaugeMetricFamily(
                   'hhvm_tc_used_bytes', 'Translation cache usage',
                   labels=['block'])
        metrics['tc'].add_metric(
                ['main'], data.get('tc-size', float('nan')))
        for block in ('hot', 'prof', 'cold', 'frozen'):
            metrics['tc'].add_metric(
                [block], data.get('tc-{}size'.format(block), float('nan')))

        for metric in metrics.values():
            yield metric

    def _collect_memory(self, data):
        if data is None:
            raise StopIteration()

        metrics = {
           'success': GaugeMetricFamily(
                   'hhvm_memory_success',
                   'HHVM was able to fetch its memory statistics'),
           'vmsize': GaugeMetricFamily(
                   'hhvm_process_memory_size_bytes',
                   'Virtual memory size of HHVM process'),
           'vmdetail': GaugeMetricFamily(
                   'hhvm_process_memory_bytes', 'Virtual memory segment size',
                   labels=['segment']),
           'strings_count': GaugeMetricFamily(
                   'hhvm_memory_strings_count', 'Static strings allocated'),
           'strings_bytes': GaugeMetricFamily(
                   'hhvm_memory_strings_bytes',
                   'Static strings total bytes used'),
        }

        memorym = data.get('Memory', {})
        metrics['success'].add_metric([], data.get('Success', 0))

        statm = memorym.get('Process Stats (bytes)', {})
        metrics['vmsize'].add_metric([], statm.get('VmSize'))
        metrics['vmdetail'].add_metric(['rss'], statm.get('VmRss'))
        metrics['vmdetail'].add_metric(['shared'], statm.get('Shared'))
        metrics['vmdetail'].add_metric(['text'], statm.get('Text(Code)'))
        metrics['vmdetail'].add_metric(['data'], statm.get('Data'))

        stringsm = memorym.get('Breakdown', {}).get('Static Strings')
        if stringsm:
            metrics['strings_count'].add_metric(
                    [], stringsm.get('Details', {}).get('Count', float('nan')))
            metrics['strings_bytes'].add_metric(
                    [], stringsm.get('Bytes', float('nan')))

        for metric in metrics.values():
            yield metric

    def _collect_status(self, data):
        if data is None:
            raise StopIteration()

        metrics = {
           'compiler': GaugeMetricFamily(
                   'hhvm_build_info', 'HHVM build information',
                   labels=['compiler', 'build']),
           'start': CounterMetricFamily(
                   'hhvm_startup', 'HHVM process start time'),
        }

        processm = data.get('status', {}).get('process', {})
        if processm:
            metrics['compiler'].add_metric([
                processm.get('compiler'),
                processm.get('build')], 1)

            hhvm_start = processm.get('start', '')
            hhvm_start_date = datetime.datetime.strptime(
                    hhvm_start, '%a, %d-%b-%Y %H:%M:%S %Z')
            metrics['start'].add_metric(
                    [], float(hhvm_start_date.strftime('%s')))

        for metric in metrics.values():
            yield metric


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument('--admin-url', metavar='URL', help='HHVM admin url',
                        default='http://localhost:9002')
    parser.add_argument('-l', '--listen', metavar='ADDRESS',
                        help='Listen on this address', default=':9192')
    parser.add_argument('-d', '--debug', action='store_true',
                        help='Enable debug logging')
    args = parser.parse_args()

    if args.debug:
        logging.basicConfig(level=logging.DEBUG)
    else:
        logging.basicConfig(level=logging.WARNING)

    address, port = args.listen.split(':', 1)

    log.info('Starting hhvm_exporter on %s:%s', address, port)

    REGISTRY.register(HHVMCollector(args.admin_url))
    start_http_server(int(port), addr=address)

    try:
        while True:
            time.sleep(1)
    except KeyboardInterrupt:
        return 1


if __name__ == "__main__":
    sys.exit(main())
